wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/AdfsCredentialsProviderFactory.java

```diff
@@ -0,0 +1,237 @@
+  private final PluginService pluginService;
+  private final TelemetryFactory telemetryFactory;
+  private final CloseableHttpClient httpClient;
+  private TelemetryContext telemetryContext;
+
+  public AdfsCredentialsProviderFactory(PluginService pluginService, CloseableHttpClient httpClient) {
+    this.pluginService = pluginService;
+    this.telemetryFactory = this.pluginService.getTelemetryFactory();
+    this.httpClient = httpClient;
+  }
+
+  @Override
+  String getSamlAssertion(final @NonNull Properties props) {
+    this.telemetryContext = telemetryFactory.openTelemetryContext(TELEMETRY_FETCH_SAML, TelemetryTraceLevel.NESTED);
+
+    String uri = "https://" + FederatedAuthConnectionPlugin.IDP_ENDPOINT.getString(props) + ':'
+        + FederatedAuthConnectionPlugin.IDP_PORT.getString(props) + "/adfs/ls/IdpInitiatedSignOn.aspx?loginToRp="
+        + FederatedAuthConnectionPlugin.RELAYING_PARTY_ID.getString(props);
+
+    try {
+      LOGGER.info(Messages.get("AdfsCredentialsProviderFactory.signOnPageUrl", new Object[] {uri}));
+      validateURL(uri);
+      HttpGet get = new HttpGet(uri);
+      CloseableHttpResponse resp = httpClient.execute(get);
+
+      if (resp.getStatusLine().getStatusCode() != 200) {
+        throw new IOException(Messages.get("AdfsCredentialsProviderFactory.signOnPageRequestFailed",
+            new Object[] {
+                resp.getStatusLine().getStatusCode(),
+                resp.getStatusLine().getReasonPhrase(),
+                EntityUtils.toString(resp.getEntity())}));
+      }
+
+      String body = EntityUtils.toString(resp.getEntity());
+
+      String action = getFormAction(body);
+      if (!StringUtils.isNullOrEmpty(action) && action.startsWith("/")) {
+        uri = "https://" + FederatedAuthConnectionPlugin.IDP_ENDPOINT.getString(props) + ':' + FederatedAuthConnectionPlugin.IDP_PORT.getString(props) + action;
+      }
+
+      LOGGER.info(Messages.get("AdfsCredentialsProviderFactory.signOnPagePostActionUrl", new Object[] {uri}));
+
+      validateURL(uri);
+      HttpPost post = new HttpPost(uri);
+      post.setEntity(new UrlEncodedFormEntity(getParameters(body, props)));
+      resp = httpClient.execute(post);
+      if (resp.getStatusLine().getStatusCode() != 200) {
+        throw new IOException(Messages.get("AdfsCredentialsProviderFactory.signOnPagePostActionRequestFailed",
+            new Object[] {
+                resp.getStatusLine().getStatusCode(),
+                resp.getStatusLine().getReasonPhrase(),
+                EntityUtils.toString(resp.getEntity())}));
+      }
+
+      String content = EntityUtils.toString(resp.getEntity());
+      Matcher matcher = FederatedAuthConnectionPlugin.SAML_PATTERN.matcher(content);
+      if (!matcher.find()) {
+        throw new IOException(Messages.get("AdfsCredentialsProviderFactory.failedLogin", new Object[] {content}));
+      }
+
+      return matcher.group(1);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```

Comment:
Do we want to add a message?

Reply:
this.telemetryContext.setSuccess(false);
this.telemetryContext.setException(e);